### PR TITLE
Feedback form - rest part

### LINKED ISSUE #2836: Form to provide feedback to the DSpace administrator
From release plan spreadsheet 

Estimate from release plan: 12 hours
Expressing interest: none

This ticket is for any necessary backend/REST API changes.  UI ticket at DSpace/dspace-angular#763

### PR DESCRIPTION
**Pull request by corrad82-4s**
## References
_Add references/links to any related issues or PRs. These may include:_
* Fixes #2836
* Related to [REST Contract](https://github.com/DSpace/RestContract/blob/main/feedbacks.md) 
* Related to https://github.com/DSpace/dspace-angular/pull/1472

## Description
This is the rest part allowing the DSpace user / visitor to send his/her feedback. Feedback is actually sent if and only if feedback.recipient property is set. An Authorization Feature performing this check is provided.

## Instructions for Reviewers

List of changes in this PR:
* A new authorization feature defining wether or not a feedback can be sent is provided (CanSendFeedbackFeature), if the above mentioned property is set, the feature returns true, false otherwise.
* A new RestRepository (FeedbackRestRepository) is provided. This repository collaborates with FeedbackService implementation (see next point) to send the feedback the DSpace user wants to send to configured recipient. If no recipient where feedback must be sent is configured, an exception is thrown in case of POST request against this repository
* A new service, FeedbackService, and its implementation are provided. Service is responsible for mail sending operation.

**Include guidance for how to test or review your PR.** 
This PR can be tested together with its [Angular counterpart](https://github.com/DSpace/dspace-angular/pull/1472), or by performing simple rest calls against authorizations endpoint to check the new authorization feature, or against /api/tools/feedbacks endpoint by posting a data structure like this
```
{
  "e-mail": "myemail@test.com",
  "message": "DSpace is the best!",
  "page": "/home"
}
```
In order to have mail actually sent, smtp must be properly configured.

## Checklist

- [X] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [X] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [X] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [X] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [X] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/content/service/FeedbackService.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import org.dspace.core.Context;
+
+/**
+ * Service interface class for the Feedback object.
+ * 
+ * @author Mykhaylo Boychuk (mykhaylo.boychuk@4science.com)
+ */
+public interface FeedbackService {
+
+    public void sendEmail(Context context, HttpServletRequest request, String recipientEmail, String senderEmail,
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/content/service/FeedbackService.java:22]: Please add JavaDocs to this method to describe what it does and what the parameters are

File: dspace-api/src/main/java/org/dspace/content/FeedbackServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import javax.mail.MessagingException;
+import javax.servlet.http.HttpServletRequest;
+
+import org.apache.commons.lang.StringUtils;
+import org.dspace.content.service.FeedbackService;
+import org.dspace.core.Context;
+import org.dspace.core.Email;
+import org.dspace.core.I18nUtil;
+
+public class FeedbackServiceImpl implements FeedbackService {
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/content/FeedbackServiceImpl.java:21]: All classes should have JavaDocs to describe them. Please add JavaDocs

File: dspace-api/src/main/java/org/dspace/content/FeedbackServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    @Override
+    public void sendEmail(Context context, HttpServletRequest request, String recipientEmail, String senderEmail,
+            String message, String page) throws IOException, MessagingException {
+        String session = request.getSession().getId();
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/content/FeedbackServiceImpl.java:26]: The backend is supposed to be stateless, which means we don't / shouldn't have a Session (or a Session ID) on the backend.  I'd recommend we replace this with the `correlationID`, as that's the "session-like" ID which we use in our dspace.log.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/authorization/impl/CanSendFeedbackFeature.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    public static final String NAME = "canSendFeedback";
+
+    @Autowired
+    private ConfigurationService configurationService;
+
+    @Override
+    @SuppressWarnings("rawtypes")
+    public boolean isAuthorized(Context context, BaseObjectRest object) throws SQLException {
+        String recipientEmail = configurationService.getProperty("feedback.recipient");
+        return StringUtils.isNotBlank(recipientEmail);
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/authorization/impl/CanSendFeedbackFeature.java:41]: This seems to be different behavior from 6.x, in that if the `feedback.recipient` is **empty** then the feedback form is disabled.  However, I like this behavior, so I'm glad you thought of it.

That said, I'd ask that we document this behavior in the `dspace.cfg` in a comment above the `feedback.recipient` setting, by saying something like:
```
# If this setting is empty or commented out, the Feedback form is disabled.
feedback.recipient = dspace-help@myu.edu
```

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/FeedbackRestRepository.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    @Override
+    @PreAuthorize("permitAll()")
+    protected FeedbackRest createAndReturn(Context context) throws AuthorizeException, SQLException {
+        HttpServletRequest req = getRequestService().getCurrentRequest().getHttpServletRequest();
+        ObjectMapper mapper = new ObjectMapper();
+        FeedbackRest feedbackRest = null;
+
+        String recipientEmail = configurationService.getProperty("feedback.recipient");
+        if (StringUtils.isBlank(recipientEmail)) {
+            throw new DSpaceFeedbackNotFoundException("Recipient's email was not found!");
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/FeedbackRestRepository.java:65]: I'd recommend we provide more descriptive error text here, perhaps "Feedback recipient is disabled. Feedback cannot be sent at this time.".  

Saying the value is "not found" implies that the request forgot to send the recipients email address....but the real problem here is that it has been disabled in the configuration.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/FeedbackRestRepository.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            feedbackRest = mapper.readValue(req.getInputStream(), FeedbackRest.class);
+        } catch (IOException exIO) {
+            throw new UnprocessableEntityException("error parsing the body " + exIO.getMessage(), exIO);
+        }
+
+        String senderEmail = feedbackRest.getEmail();
+        String message = feedbackRest.getMessage();
+
+        if (StringUtils.isBlank(senderEmail) || StringUtils.isBlank(message)) {
+            throw new DSpaceBadRequestException("Filds as e-mail and message are mandatory!");
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/FeedbackRestRepository.java:78]: Small spelling/grammar errors here. Should say something like "The fields email and message are mandatory!"

**Review by tdonohue — CHANGES_REQUESTED** [review]
@corrad82-4s & @Micheleboychuk : Great work here! I tested this with the frontend PR (https://github.com/DSpace/dspace-angular/pull/1472) and overall this works well and looks good. However, there's some minor code cleanup that I think needs to be done before this is ready to merge.  None of this should be major though, and overall it doesn't impact the behavior of the current code.

**Comment by corrad82-4s** [comment]
hi @tdonohue , thank you for your feedback. I have updated this PR with suggested changes

**Review by tdonohue — APPROVED** [review]
👍 Re-tested today with the updated UI PR (https://github.com/DSpace/dspace-angular/pull/1472).  This all looks good and works well. Thanks @corrad82-4s  and @Micheleboychuk !  Will merge this as soon as the frontend PR approvals are completed.

**Pull request merged**